### ISSUE TITLE
Support deploying multiple contracts in parallel

### ISSUE THREAD

**Issue by Arachnid** [wighawag/hardhat-deploy#154]
hardhat-deploy works fine if you `await` each `deploy` call before calling the next - but if you try and queue up several and `await` them all at the end, it appears it tries to reuse the same nonce for each transaction, resulting in `REPLACEMENT_UNDERPRICED` errors. It'd be nice if it handled this correctly so we can speed up deploys by deploying contracts in parallel where practical.

**Comment by wighawag** [comment]
Thanks for your report, I actually recently changed the behaviour  to not use the pending nonce for a particular reason : ability to restart the process and ensuring no new tx get triggered (this was an issue that users reported with the use of pending state, especially when the first process was using a low gas price and so the use of pending nonce was forcing the script to keep creating a new tx on top of the pending one)

But if your use case differs, I would be happy to add an option, I guess this would fit more as a global option ?

**Comment by Arachnid** [comment]
Hm, I think at a minimum it needs to handle not `await`ing calls immediately without failing. The reason this is happening is probably because I'm sending a second tx before the first one has been received by the node - Ethers handles this by making you `await` the transaction request, and then `await` the transaction itself to wait for it to be mined. It'd be nice to be able to do something similar here.

**Comment by wighawag** [comment]
I think the error would still appear regardless of ordering issue since it use the confirmed nonce for the reason I mentioned above.

The deploy function is an helper function that do more than just performing the tx and allowing it to be done in 2 phase will change how hardhat-deploy deploy work and prevent some features like the ability to reexecute the whole deploy scripts with peace in mind.

Maybe you are better of using ethers factory and call hardhat-deploy `deployments.save` method to keep track of deployment.

I ll think about it more, currently not much available though.

**Comment by wighawag** [comment]
Hey @Arachnid , forgot about one option that might suits your use case, when you execute deploy you can set the `nonce` option to be `"pending"` this will use the pending nonce instead of the confirmed one.

```js
deploy("AContract", {from: deployer, nonce: "pending"});
deploy("BContract", {from: deployer, nonce: "pending"});
```

A consequence of this though, as mentioned above, is that if you have one of these tx stuck in the mempool, and you re-execute  the script and decide to skip waiting, this, will use a further nonce, making the script not 100% idempotent.

**Comment by wighawag** [comment]
Closing this for now, feel free to reopen or create an new issue if the current behavior could be improved
